File: java/08_054.java
import java.util.*;

class TreeNode {
   int val;
   TreeNode left;
   TreeNode right;
   
   TreeNode(int x) {
      val = x;
   }
}

public class Solution {
   private Map<String, Integer> memo;
   
   /**
    * Find minimum possible height after removing k nodes
    */
   public int minHeightAfterRemoval(TreeNode root, int k) {
      memo = new HashMap<>();
      return dfs(root, k);
   }
   
   /**
    * DFS with memoization to explore removal options
    */
   private int dfs(TreeNode node, int remainingRemovals) {
      if (node == null) {
         return 0;
      }
      
      // Check if this state is already computed
      String key = node.toString() + ":" + remainingRemovals;
      if (memo.containsKey(key)) {
         return memo.get(key);
      }
      
      // Option 1: Remove this node (height becomes 0)
      int heightIfRemoved = 0;
      
      // Option 2: Keep this node
      int leftHeight = dfs(node.left, remainingRemovals);
      int rightHeight = dfs(node.right, remainingRemovals);
      int heightIfKept = 1 + Math.max(leftHeight, rightHeight);
      
      int minHeight = heightIfKept;
      
      // Try removing the node if possible
      if (remainingRemovals > 0) {
         minHeight = Math.min(minHeight, heightIfRemoved);
      }
      
      // Try removing subtrees in different combinations
      if (remainingRemovals > 0) {
         for (int removeLeft = 0; removeLeft <= remainingRemovals; 
               removeLeft++) {
            int removeRight = remainingRemovals - removeLeft;
            int heightRemoveSubtrees = 1 + Math.max(
               dfs(node.left, removeLeft),
               dfs(node.right, removeRight)
            );
            minHeight = Math.min(minHeight, heightRemoveSubtrees);
         }
      }
      
      // Cache result
      memo.put(key, minHeight);
      return minHeight;
   }
}

File: data structures and algorithms/cs/08_054.cs
using System.Collections.Generic;

public class TreeNode
{
   public int Val { get; set; }
   public TreeNode Left { get; set; }
   public TreeNode Right { get; set; }

   public TreeNode(int x)
   {
      Val = x;
   }
}

public class Solution
{
   private Dictionary<string, int> memo;

   public int MinHeightAfterRemoval(
      TreeNode root, int k)
   {
      memo = new Dictionary<string, int>();
      return Dfs(root, k);
   }

   private int Dfs(TreeNode node,
      int remainingRemovals)
   {
      if (node == null)
      {
         return 0;
      }

      string key = node.ToString() +
         ":" + remainingRemovals;
      if (memo.ContainsKey(key))
      {
         return memo[key];
      }

      int heightIfRemoved = 0;
      int leftHeight = Dfs(node.Left,
         remainingRemovals);
      int rightHeight = Dfs(node.Right,
         remainingRemovals);
      int heightIfKept = 1 + Math.Max(
         leftHeight, rightHeight);

      int minHeight = heightIfKept;

      if (remainingRemovals > 0)
      {
         minHeight = Math.Min(minHeight,
            heightIfRemoved);
      }

      if (remainingRemovals > 0)
      {
         for (int removeLeft = 0;
            removeLeft <= remainingRemovals;
            removeLeft++)
         {
            int removeRight = remainingRemovals -
               removeLeft;
            int heightRemoveSubtrees = 1 + Math.Max(
               Dfs(node.Left, removeLeft),
               Dfs(node.Right, removeRight)
            );
            minHeight = Math.Min(
               minHeight,
               heightRemoveSubtrees);
         }
      }

      memo[key] = minHeight;
      return minHeight;
   }
}